rust/xaynet-core/src/mask/scalar.rs

```diff
@@ -0,0 +1,250 @@
+//! Scalar representation and conversion.
+//!
+//! See the [mask module] documentation since this is a private module anyways.
+//!
+//! [mask module]: ../index.html
+
+use crate::mask::{
+    model::{ratio_to_float, PrimitiveType},
+    PrimitiveCastError,
+};
+use derive_more::{From, Into};
+use num::{
+    clamp,
+    rational::Ratio,
+    traits::{float::FloatCore, ToPrimitive},
+    BigInt,
+    BigUint,
+    One,
+    Unsigned,
+    Zero,
+};
+use serde::{Deserialize, Serialize};
+use std::{convert::TryFrom, fmt::Debug};
+use thiserror::Error;
+
+#[derive(Debug, Clone, PartialEq, Hash, From, Into, Serialize, Deserialize)]
+/// A numerical representation of a machine learning scalar.
+pub struct Scalar(Ratio<BigUint>);
+
+impl Scalar {
+    /// Constructs a new `Scalar` from the given numerator and denominator.
+    pub fn new<U>(numer: U, denom: U) -> Self
+    where
+        U: Unsigned + Into<BigUint>,
+    {
+        Self(Ratio::new(numer.into(), denom.into()))
+    }
+
+    /// Constructs a `Scalar` of unit value.
+    pub fn unit() -> Self {
+        Self(Ratio::one())
+    }
+
+    /// Convenience method for conversion into a non-negative ratio of `BigInt`.
+    pub(crate) fn into_ratio(self) -> Ratio<BigInt> {
+        let (numer, denom) = self.0.into();
+        Ratio::new(numer.into(), denom.into())
+    }
+
+    /// Convenience method for conversion to a non-negative ratio of `BigInt`.
+    pub(crate) fn to_ratio(&self) -> Ratio<BigInt> {
+        let numer = self.0.numer().clone();
+        let denom = self.0.denom().clone();
+        Ratio::new(numer.into(), denom.into())
+    }
+
+    fn from_ratio(r: Ratio<BigInt>) -> anyhow::Result<Self> {
+        let (inumer, idenom) = r.into();
+        let unumer = BigUint::try_from(inumer)?;
+        let udenom = BigUint::try_from(idenom)?;
+        Ok(Self(Ratio::new(unumer, udenom)))
+    }
+}
+
+#[derive(Error, Debug)]
+#[error("Could not convert weight {weight} to primitive type {target}")]
+/// Errors related to scalar conversion into primitives.
+pub struct ScalarCastError {
+    weight: Ratio<BigUint>,
+    target: PrimitiveType,
+}
+
+/// An interface for conversion into a primitive value.
+///
+/// This trait is used to convert a [`Scalar`], which has its own internal
+/// representation, into a primitive type ([`f32`], [`f64`], [`i32`], [`i64`]).
+/// The opposite trait is [`FromPrimitive`].
+pub trait IntoPrimitive<P>: Sized {
```

Comment:
as a side note, we should probably later refactor the `IntoPrimitive` and `FromPrimitive` traits, not only for `Scalar` but also for `Model`.

for example the conversion error gets always constructed because of `.ok_or()`, where we should use `.ok_or_else()` for lazy evaluation.

also we have a duplication of trait declarations now, which only differ in the output type. we might want to unify them to something similar to
```rust
pub trait IntoPrimitives<P>: Sized {
    type Output<P>;

    fn into_primitive(self) -> Self::Output<P>;

    fn to_primitive(&self) -> Self::Output<P>;

    fn into_primitive_unchecked(self) -> P;
}
```

Reply:
>we might want to unify them to something similar to ...

this is a very nice idea. while we're at it, we can also lift some types up from `model` that are also used by `scalar`, like `PrimitiveCastError` and `PrimitiveType`. I will write a ticket.

>the conversion error gets always constructed because of `.ok_or()`, where we should use `.ok_or_else()` for lazy evaluation.

I also had this idea, but Clippy thought otherwise: https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_lazy_evaluations. although strangely, it doesn't complain about my use of `unwrap_or_else` e.g.
```rust
fn from_primitive_bounded(prim: i32) -> Self {
    Self::from_primitive(prim).unwrap_or_else(|_| Self(Ratio::zero()))
}
```